src/Grpc.Net.Client/Balancer/Internal/SocketConnectivitySubchannelTransport.cs

```diff
@@ -199,77 +207,144 @@ public async ValueTask<ConnectResult> TryConnectAsync(ConnectContext context)
+                                _initialSocketData.Add(buffer.AsMemory(0, readCount));
+                            }
+                            else
+                            {
+                                hasReadData = false;
+                            }
+                        }
+                        while (hasReadData);
+                    }
+                    catch (Exception ex)
+                    {
+                        closeSocket = true;
+                        checkException = ex;
+                        SocketConnectivitySubchannelTransportLog.ErrorCheckingSocket(_logger, _subchannel.Id, socketAddress, ex);
+                    }
                 }
+            }
 
-                if (closeSocket)
+            if (closeSocket)
+            {
+                lock (Lock)
                 {
-                    lock (Lock)
+                    if (_disposed)
                     {
-                        if (_disposed)
-                        {
-                            return;
-                        }
+                        return;
+                    }
 
-                        if (_initialSocket == socket)
-                        {
-                            DisconnectUnsynchronized();
-                        }
+                    if (_initialSocket == socket)
+                    {
+                        DisconnectUnsynchronized();
                     }
-                    _subchannel.UpdateConnectivityState(ConnectivityState.Idle, new Status(StatusCode.Unavailable, "Lost connection to socket.", sendException));
                 }
+                _subchannel.UpdateConnectivityState(ConnectivityState.Idle, new Status(StatusCode.Unavailable, "Lost connection to socket.", checkException));
             }
         }
         catch (Exception ex)
         {
             SocketConnectivitySubchannelTransportLog.ErrorSocketTimer(_logger, _subchannel.Id, ex);
         }
+
+        lock (Lock)
+        {
+            if (!_disposed)
+            {
+                // Schedule next ping.
+                _socketConnectedTimer.Change(_socketPingInterval, Timeout.InfiniteTimeSpan);
+            }
+        }
     }
 
     public async ValueTask<Stream> GetStreamAsync(BalancerAddress address, CancellationToken cancellationToken)
     {
         SocketConnectivitySubchannelTransportLog.CreatingStream(_logger, _subchannel.Id, address);
 
         Socket? socket = null;
+        BalancerAddress? socketAddress = null;
+        List<ReadOnlyMemory<byte>>? socketData = null;
         lock (Lock)
         {
             if (_initialSocket != null)
             {
                 var socketAddressMatch = Equals(_initialSocketAddress, address);
 
                 socket = _initialSocket;
+                socketAddress = _initialSocketAddress;
+                socketData = _initialSocketData;
                 _initialSocket = null;
```

Comment:
add  **_socketConnectedTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);**  
because _initialSocket  always null,until the next call TryConnectAsync  assign .